File: graphics/3d/4/victor_wu/parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "ml6.h"
#include "display.h"
#include "draw.h"
#include "matrix.h"
#include "parser.h"


/*======== void parse_file () ==========
Inputs:   char * filename 
          struct matrix * transform, 
          struct matrix * pm,
          screen s
Returns: 

Goes through the file named filename and performs all of the actions listed in that file.
The file follows the following format:
     Every command is a single character that takes up a line
     Any command that requires arguments must have those arguments in the second line.
     The commands are as follows:
         l: add a line to the edge matrix - 
	    takes 6 arguemnts (x0, y0, z0, x1, y1, z1)
	 c: add a circle to the edge matrix - 
	    takes 3 arguments (cx, cy, r)
	 h: add a hermite curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	 b: add a bezier curve to the edge matrix -
	    takes 8 arguments (x0, y0, x1, y1, x2, y2, x3, y3)
	 i: set the transform matrix to the identity matrix - 
	 s: create a scale matrix, 
	    then multiply the transform matrix by the scale matrix - 
	    takes 3 arguments (sx, sy, sz)
	 t: create a translation matrix, 
	    then multiply the transform matrix by the translation matrix - 
	    takes 3 arguments (tx, ty, tz)
	 x: create an x-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 y: create an y-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 z: create an z-axis rotation matrix,
	    then multiply the transform matrix by the rotation matrix -
	    takes 1 argument (theta)
	 a: apply the current transformation matrix to the 
	    edge matrix
	 v: draw the lines of the edge matrix to the screen
	    display the screen
	 g: draw the lines of the edge matrix to the screen
	    save the screen to a file -
	    takes 1 argument (file name)
	 q: end parsing

See the file script for an example of the file format


IMPORTANT MATH NOTE:
the trig functions int math.h use radian mesure, but us normal
humans use degrees, so the file will contain degrees for rotations,
be sure to conver those degrees to radians (M_PI is the constant
for PI)

03/08/12 16:22:10
jdyrlandweaver
====================*/
void parse_file ( char * filename, 
                  struct matrix * transform, 
                  struct matrix * pm,
                  screen s) {

  FILE *f;
  char line[256];
  struct matrix * tmp;
  double angle;
  color g;

  g.red = 255;
  g.green = 0;
  g.blue = 255;
  
  clear_screen(s);

  if ( strcmp(filename, "stdin") == 0 ) 
    f = stdin;
  else
    f = fopen(filename, "r");
  
  while ( fgets(line, 255, f) != NULL ) {
    line[strlen(line)-1]='\0';
    //printf(":%s:\n",line);
    char c;
    double x, y, z, x1, y1, z1;
    //, x2, y2, z2, x3, y3, z3, x4, y4, z4;
    double r;
    c = line[0];
    
    switch (c) {
    case 'p':
      fgets(line, 255, f);
      double w, h, d;
      sscanf(line, "%lf %lf %lf %lf %lf %lf", &x, &y, &z, &w, &h, &d);
      add_prism(pm, x, y, z, w, h, d);
      break;
    case 'm':
      fgets(line, 255, f);
      sscanf(line, "%lf %lf %lf", &x, &y, &r);
      add_sphere(pm, x, y, 1, r, .5);
      break;
    case 'd':
      fgets(line, 255, f);
      double r1, r2;
      sscanf(line, "%lf %lf %lf %lf", &x, &y, &r1, &r2);
      add_torus(pm, x, y, 1, r1, r2, .5);
      break;
    case 'l':
      //      printf("LINE!\n");
      fgets(line, 255, f);
      //      printf("\t%s", line);
      //line[strlen(line)-1]='\0';
      sscanf(line, "%lf %lf %lf %lf %lf %lf", &x, &y, &z, &x1, &y1, &z1);
      add_edge(pm, x, y, z, x1, y1, z1);
      // printf( "%lf %lf %lf %lf %lf %lf\n", x, y, z, x1, y1, z1);
      break;
    case 'c':
      fgets(line, 255, f);
      sscanf(line, "%lf %lf %lf", &x, &y, &r);
      add_circle(pm, x, y, r, STEP);
      break;
    case 's':
      //printf("SCALE\n");
      fgets(line, 255, f);
      //line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf", &x, &y, &z);
      tmp = make_scale(x, y, z);
      matrix_mult(tmp, transform);
      //print_matrix(transform);
      break;
    case 't':
      //printf("TRANSLATE\n");
      fgets(line, 255, f);
      //      line[strlen(line)-1]='\0';      
      sscanf(line, "%lf %lf %lf", &x, &y, &z);
      tmp = make_translate(x, y, z);
      matrix_mult(tmp, transform);
      //print_matrix(transform);
      break;
    case 'x':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI / 180);
      tmp = make_rotX( angle);
      matrix_mult(tmp, transform);
      break;
    case 'y':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI / 180);
      tmp = make_rotY( angle);
      matrix_mult(tmp, transform);
      break;
    case 'z':
      //printf("ROTATE!\n");
      fgets(line, 255, f);
      sscanf(line, "%lf", &angle);
      angle = angle * (M_PI / 180);
      tmp = make_rotZ( angle);
      matrix_mult(tmp, transform);
      break;
    case 'i':
      ident(transform);
      break;
    case 'a':
      //printf("APPLY!\n");
      //print_matrix( transform );
      //      print_matrix(pm);
      matrix_mult(transform, pm);
      break;
    case 'v':
      clear_screen(s);
      draw_lines(pm, s, g);
      display(s);
      break;
    case 'g':
      fgets(line, 255, f);
      // line[strlen(line)-1] = '\0';
      clear_screen(s);
      draw_lines(pm, s, g);
      save_extension(s, line);
      break;
    case 'q':
      return;
    default:
      printf("Invalid command\n");
      break;
    }
  }

  free_matrix(tmp);
  fclose(f);
  //printf("END PARSE\n");
}

  

File: graphics/3d/4/alex_libman/parser.py
from display import *
from matrix import *
from draw import *

color = [0, 255, 0]

def parse_file( fname, points, transform ):
    f = open(fname,"r")
    lines = f.readlines()
    f.close()
    screen = new_screen()
    i = 0
    while i < len(lines):
        l = lines[i].replace("\n","")
        if l == "l":
            coor = lines[i+1].replace("\n","").split(" ")
            coor = [float(q) for q in coor]
            add_edge(points, coor[0], coor[1], coor[2], coor[3], coor[4], coor[5])
            
        elif l == "c":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_circle(points, values[0], values[1], 0, values[2], 0.01)
        elif l == "h":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_curve(points,values[0],values[1],values[2],values[3],values[4],
                      values[5],values[6],values[7],0.01,"hermite")
        elif l == "b":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_curve(points,values[0],values[1],values[2],values[3],values[4],
                      values[5],values[6],values[7],0.01,"bezier")

        elif l == "p":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_prism(points,values[0],values[1],values[2],values[3],values[4],values[5])
        elif l == "m":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_sphere(points,values[0],values[1],values[2],0.02)
        elif l == "d":
            values = [float(q) for q in lines[i+1].replace("\n","").split(" ")]
            add_torus(points,values[0],values[1],values[2],values[3],0.02)
            
        elif l == "i":
            ident(transform)
        elif l == "s":
            values = lines[i+1].replace("\n","").split(" ")
            values = [float(q) for q in values]
            matrix_mult(make_scale(values[0], values[1], values[2]),transform)
        elif l == "t":
            values = lines[i+1].replace("\n","").split(" ")
            values = [float(q) for q in values]
            matrix_mult(make_translate(values[0], values[1], values[2]),transform)
        elif l == "x":
            theta = float(lines[i+1].replace("\n",""))
            matrix_mult(make_rotX(theta),transform)
        elif l == "y":
            theta = float(lines[i+1].replace("\n",""))
            matrix_mult(make_rotY(theta),transform)
        elif l == "z":
            theta = float(lines[i+1].replace("\n",""))
            matrix_mult(make_rotZ(theta),transform)
        elif l == "a":
            matrix_mult(transform,points)
            
        elif l == "w":
            points = []
        elif l == "vn":
            draw_lines(points, screen, color)
        elif l == "v":
            draw_lines(points, screen, color)
            display(screen)
        elif l == "g":
            fname = lines[i+1].replace("\n","")
            draw_lines(points, screen, color)
            save_ppm(screen, fname)
        i += 1
        
         

points = []
transform = new_matrix()

parse_file( 'script_c', points, transform )
